assets/js/ConnectFour/Game.js

```diff
@@ -84,54 +119,112 @@ customElements.define('connect-four-game', class extends HTMLElement {
                 .classList
-                .add('gp-game__field--highlight')
+                .add('gp-game__field--highlight'), Math.random() * 100)
             );
         });
     }
 
     /**
-     * Only show if the user follows the moves. Otherwise, notify user that a new move is available.
-     *
-     * @param {import('./Model/Game.js').Move} move
+     * @param {Number} column
+     * @returns {HTMLElement|undefined}
      */
-    _onMoveAppendedToGame(move) {
-        if (this._followMovesButton.disabled === true) {
-            this._showMove(move);
-            this._numberOfCurrentMoveInView++;
-            this._updateNavigationButtons();
-            this._showWinningSequences();
-        } else {
-            this._followMovesButton.classList.add('btn-warning', 'icon-tada');
-        }
+    _lastFieldInColumn(column) {
+        const fields = this._gameNode.querySelectorAll(
+            `.gp-game__field[data-column="${column}"]:not(.gp-game__field--red):not(.gp-game__field--yellow)`
+        );
+        return fields[fields.length - 1];
+    }
+
+    _removeFieldPreview() {
+        this._gameNode.querySelector(`.${this._previewClass()}`)?.classList.remove(this._previewClass());
     }
 
     _onFieldClick(event) {
-        let cell = event.target;
+        if (this._isMoveInProgress) return;
+
+        this._isMoveInProgress = true;
 
-        let loadingTimeout = setTimeout(() => this._gameNode.classList.add('gp-loading'), 250);
+        const field = this._lastFieldInColumn(event.target.dataset.column);
+        if (!field) {
+            this._isMoveInProgress = false;
+            return;
+        }
 
-        service.move(this._game.gameId, cell.dataset.column)
-            .catch(() => true)
-            .finally(() => {
-                if (loadingTimeout) clearTimeout(loadingTimeout);
-                this._gameNode.classList.remove('gp-loading');
-            });
+        const eventOptions = {
+            bubbles: true,
+            detail: {
+                x: parseInt(field.dataset.column),
+                y: parseInt(field.dataset.row),
+                color: this._playerId === this._game.redPlayerId ? 1 : 2,
+                playerId: this._playerId,
+                nextPlayerId: this._playerId === this._game.redPlayerId
+                    ? this._game.yellowPlayerId
+                    : this._game.redPlayerId,
+                preview: true
+            }
+        };
+        this.dispatchEvent(new CustomEvent('ConnectFour.PlayerMoved', eventOptions));
+        this._removeFieldPreview();
+
+        const numberOfMoves = this._game.numberOfMoves();
+        service.move(this._game.gameId, field.dataset.column)
+            .catch(() => {
+                if (numberOfMoves !== this._game.numberOfMoves()) return;
+
+                this.dispatchEvent(new CustomEvent('ConnectFour.PlayerMovedFailed', eventOptions));
```

Comment:
When the move comes through SSE, but the request fails at the end, e.g. for some code that gets executed after the domain operation has been done, then the move will rollback in the UI. That case should be accounted for.